### PR TITLE
RavenDB-13659 Possible deadlock when changing rachis state and execut…

### PR DESCRIPTION
**Pull request by karmeli87**
…ing `SwitchToCandidateStateOnTimeout`

### CODE REVIEW

File: src/Raven.Server/Rachis/TimeoutEvent.cs [hunk 1]
```diff
@@ -117,27 +117,32 @@ public void DisableTimeout()
             }
         }
 
+        private int _inProgress;
+
         public void ExecuteTimeoutBehavior()
         {
-            lock (this)
-            {
-                if (Disable)
```
Comment (ppekrol) [inline src/Raven.Server/Rachis/TimeoutEvent.cs:124]: Shouldn't we take into account that it might be disabled?

Comment (karmeli87) [reply src/Raven.Server/Rachis/TimeoutEvent.cs:124]: yes, was too eager on removing the `lock` :)

File: src/Raven.Server/Rachis/TimeoutEvent.cs [hunk 2]
```diff
@@ -117,27 +117,41 @@ public void DisableTimeout()
             }
         }
 
+        private int _inProgress;
+
         public void ExecuteTimeoutBehavior()
         {
+            var action = _timeoutHappened;
+
+            if (Interlocked.CompareExchange(ref _inProgress, 1, 0) == 1)
```
Comment (ayende) [inline src/Raven.Server/Rachis/TimeoutEvent.cs:124]: What is the dead lock that you are running into here?
And why not handle that using the already existing `lock (this)` by changing this to be `Monitor.TryEnter` ?

Comment (karmeli87) [reply src/Raven.Server/Rachis/TimeoutEvent.cs:124]: p1 is the timer callback (`SwitchToCandidateStateOnTimeout`)
p2 switching the rachis state (`SetNewState`)

p1 enters the lock -> p2 opens write tx -> p2 try to enter the lock but blocked -> p1 tries to open write tx.

Comment (karmeli87) [reply src/Raven.Server/Rachis/TimeoutEvent.cs:124]: Not sure how `Monitor.TryEnter` would help.
I need that the execution of the callback to be outside of the lock.

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Rachis/TimeoutEvent.cs [hunk 3]
```diff
@@ -117,27 +117,41 @@ public void DisableTimeout()
...
 
+        private int _inProgress;
+
         public void ExecuteTimeoutBehavior()
         {
-            lock (this)
-            {
-                if (Disable)
-                    return;
+            var action = _timeoutHappened;
```
Comment (ayende) [inline src/Raven.Server/Rachis/TimeoutEvent.cs]: I don't like this.
We need to do this, but we have to capture the `_timouteHappened` value _inside_ the lock

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Rachis/TimeoutEvent.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/Rachis/TimeoutEvent.cs]: done

**Review by ayende — APPROVED** [review]

**Comment by ppekrol** [comment]
@karmeli87 we are getting a lot of timeouts in the tests results, can you check that?

**Pull request merged**